File: Runtime/STSimMapBase.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports SyncroSim.Core

MustInherit Class STSimMapBase

    Private m_Scenario As Scenario
    Private m_PrimaryStratumLabel As String
    Private m_SecondaryStratumLabel As String
    Private m_HasItems As Boolean

    Protected Sub New(ByVal scenario As Scenario)

        Me.m_Scenario = scenario

        Dim ds As DataSheet = scenario.Project.GetDataSheet(DATASHEET_TERMINOLOGY_NAME)
        GetStratumLabelTerminology(ds, Me.m_PrimaryStratumLabel, Me.m_SecondaryStratumLabel)

    End Sub

    Protected ReadOnly Property PrimaryStratumLabel As String
        Get
            Return Me.m_PrimaryStratumLabel
        End Get
    End Property

    Protected ReadOnly Property SecondaryStratumLabel As String
        Get
            Return Me.m_SecondaryStratumLabel
        End Get
    End Property

    Protected Sub SetHasItems()
        Me.m_HasItems = True
    End Sub

    Public ReadOnly Property HasItems As Boolean
        Get
            Return Me.m_HasItems
        End Get
    End Property

    Protected Shared Sub ThrowDuplicateItemException()
        Throw New STSimMapDuplicateItemException("An item with the same keys has already been added.")
    End Sub

    Protected Shared Function FormatValue(ByVal value As Nullable(Of Integer)) As String

        If (Not value.HasValue) Then
            Return "NULL"
        Else
            Return CStr(value)
        End If

    End Function

    Protected Function GetStratumName(ByVal id As Nullable(Of Integer)) As String
        Return Me.GetDefinitionName(DATASHEET_STRATA_NAME, id)
    End Function

    Protected Function GetSecondaryStratumName(ByVal id As Nullable(Of Integer)) As String
        Return Me.GetDefinitionName(DATASHEET_SECONDARY_STRATA_NAME, id)
    End Function

    Protected Function GetStateClassName(ByVal id As Nullable(Of Integer)) As String
        Return Me.GetDefinitionName(DATASHEET_STATECLASS_NAME, id)
    End Function

    Protected Function GetTransitionGroupName(ByVal id As Nullable(Of Integer)) As String
        Return Me.GetDefinitionName(DATASHEET_TRANSITION_GROUP_NAME, id)
    End Function

    Protected Function GetTransitionTypeName(ByVal id As Nullable(Of Integer)) As String
        Return Me.GetDefinitionName(DATASHEET_TRANSITION_TYPE_NAME, id)
    End Function

    Protected Function GetTransitionAttributeTypeName(ByVal id As Nullable(Of Integer)) As String
        Return Me.GetDefinitionName(DATASHEET_TRANSITION_ATTRIBUTE_TYPE_NAME, id)
    End Function

    Protected Function GetDefinitionName(ByVal dataSheetName As String, ByVal id As Nullable(Of Integer)) As String

        If (Not id.HasValue) Then
            Return "NULL"
        Else

            Dim ds As DataSheet = Me.m_Scenario.Project.GetDataSheet(dataSheetName)
            Return ds.ValidationTable.GetDisplayName(id.Value)

        End If

    End Function

End Class

File: src/Runtime/STSimMapBase.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2023 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using SyncroSim.Core;
using System.Globalization;

namespace SyncroSim.STSim
{
    internal abstract class STSimMapBase
    {
        private Scenario m_Scenario;
        private string m_PrimaryStratumLabel;
        private string m_SecondaryStratumLabel;
        private string m_TertiaryStratumLabel;
        private bool m_HasItems;

        protected STSimMapBase(Scenario scenario)
        {
            this.m_Scenario = scenario;

            DataSheet ds = scenario.Project.GetDataSheet(Strings.DATASHEET_TERMINOLOGY_NAME);

            TerminologyUtilities.GetStratumLabelTerminology(
                ds, 
                ref this.m_PrimaryStratumLabel, 
                ref this.m_SecondaryStratumLabel, 
                ref this.m_TertiaryStratumLabel);
        }

        protected string PrimaryStratumLabel
        {
            get
            {
                return this.m_PrimaryStratumLabel;
            }
        }

        protected string SecondaryStratumLabel
        {
            get
            {
                return this.m_SecondaryStratumLabel;
            }
        }

        protected string TertiaryStratumLabel
        {
            get
            {
                return this.m_TertiaryStratumLabel;
            }
        }

        protected void SetHasItems()
        {
            this.m_HasItems = true;
        }

        public bool HasItems
        {
            get
            {
                return this.m_HasItems;
            }
        }

        public Scenario Scenario
        {
            get
            {
                return this.m_Scenario;
            }
        }

        protected static void ThrowDuplicateItemException()
        {
            throw new STSimMapDuplicateItemException(
                "An item with the same keys has already been added.");
        }

        protected static string FormatValue(int? value)
        {
            if (!value.HasValue)
            {
                return "NULL";
            }
            else
            {
                return Convert.ToString(value, CultureInfo.InvariantCulture);
            }
        }

        protected string GetStratumName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_STRATA_NAME, id);
        }

        protected string GetSecondaryStratumName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_SECONDARY_STRATA_NAME, id);
        }

        protected string GetTertiaryStratumName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_TERTIARY_STRATA_NAME, id);
        }

        protected string GetStateClassName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_STATECLASS_NAME, id);
        }

        protected string GetTransitionGroupName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_TRANSITION_GROUP_NAME, id);
        }

        protected string GetTransitionTypeName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_TRANSITION_TYPE_NAME, id);
        }

        protected string GetStateAttributeTypeName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_NAME, id);
        }

        protected string GetTransitionAttributeTypeName(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_TRANSITION_ATTRIBUTE_TYPE_NAME, id);
        }

        protected string GetTSTGroupString(int? id)
        {
            return this.GetProjectItemName(Strings.DATASHEET_TRANSITION_GROUP_NAME, id);
        }

        protected string GetProjectItemName(string dataSheetName, int? id)
        {
            if (!id.HasValue)
            {
                return "NULL";
            }
            else
            {
                DataSheet ds = this.m_Scenario.Project.GetDataSheet(dataSheetName);
                return ds.ValidationTable.GetDisplayName(id.Value);
            }
        }
    }
}
